### ISSUE TITLE
torch.nn.DataParallel slowdown dataloading due to float to half cast on CPU

### ISSUE THREAD

**Issue by h6197627** [NVIDIA/apex#152]
In case of multi-GPU training with torch.nn.DataParallel it takes CPU tensor, scatters and then transfers to GPUs. In case it is wrapped with FP16Model it will perform float to half casting on the CPU and then will move half tensor to GPUs. However according to my timings **float -> cuda -> half** is around 10 times faster then **float -> half -> cuda**
I think it can be handled in FP16Model with something like
```
if isinstance(network, nn.DataParallel):
    name, model = next(network.named_children())
    network.add_module(name, nn.Sequential(tofp16(), model))
else:
    network = nn.Sequential(tofp16(), network)
```
and deleting casting in forward method. That way casting to half will be performed on each GPU independently. Tuple handling can be moved to tofp16 class forward method.
If it is acceptable I can do pull request

**Comment by mcarilli** [comment]
We strongly prefer DistributedDataParallel (multiprocess) over DataParallel, and we design the utilities to work best with DistributedDataParallel.  I don't want to start special casing things to support DataParallel.

That being said, you're definitely right that casting to half on the CPU takes much longer.  Isn't the easiest solution in this case to either call `.cuda().half()` on your data tensor before passing it to the model?  Any additional calls to `.cuda()` or `.half()` within your network will become no-ops, and (I believe) DataParallel will scatter the tensors to the participating devices properly.  (The problem with this approach is that the entire batch will be converted to half on device 0 before being scattered, which is an imbalance, but may not be a bottleneck for your application).

Another solution, I think, is to wrap your model in FP16_Model **before** wrapping it in DataParallel, and sending in cpu-side input.  I THINK (correct me if I'm wrong) that everything will then occur the way you hope:  DataParallel's forward wrapper will scatter the inputs from the CPU to the various devices as float, then each replicated FP16_Model's forward method will be called, and cast the input chunks (which are now on each device) from float to half.

In general, I find it's a good idea to wrap with DistributedDataParallel or DataParallel only after all other modfications to your model are finished, otherwise you can run into weird gotchas.

**Comment by h6197627** [comment]
Hi @mcarilli,
thanks, I think the second solution actually looks really good for me!

Just couple of short question before closing:
1. Does apex.DistributedDataParallel provides any benefit over torch.DistributedDataParallel for ordinary user case? Or, with release of PyTorch 1.0 in majority of tasks it roughly the same?
2. While initilizing models in processes of DistributedDataParallel does initialization consistency is user's responsibility (through fixed seed, probably)? Or DistributedDataParallel handles this for user and all distributed processes get the same intial weights by default?

**Comment by mcarilli** [comment]
I maintain `apex.parallel.DistributedDataParallel` to support our internal use cases, which means it can handle some weird things that `torch.nn.parallel.DistributedDataParallel` currently doesn't.    However, for most use cases apex and torch DDP should be roughly equivalent.  I'd like to merge some features of apex DDP into torch DDP eventually.

Both apex and torch DDP broadcast the model params from rank 0 on construction, so the model params are all initialized with the same weights.  Also the params are updated with bitwise identical (allreduced) gradients in each process, so they should remain bitwise identical across ranks during training.

However, you may find that batchnorm buffers (ie running stats) may NOT remain the same across processes, because each process sees a different sequence of data.  This is just something to watch out for.  In practice we find that slightly-different batchnorm stats across processes tends not to impair accuracy.  It also does not cause the model params themselves to diverge (the params should always remain bitwise identical across processes, as long as each process calls the same update step).

**Comment by h6197627** [comment]
@mcarilli, thanks for clarifying!
